src/basic/terminal-util.c

```diff
@@ -228,49 +231,261 @@ int ask_char(char *ret, const char *replies, const char *fmt, ...) {
+                                goto fail;
+                        }
+                        if (IN_SET(cr, COMPLETION_PARTIAL, COMPLETION_FULL)) {
+                                /* Output the new suffix we learned */
+                                fputs(ASSERT_PTR(startswith(new_string, strempty(string))), stdout);
+
+                                /* And update the whole string */
+                                free_and_replace(string, new_string);
+                                n = strlen(string);
+                        }
+                        if (cr == COMPLETION_NONE)
+                                fputc('\a', stdout); /* BEL */
+
+                        if (IN_SET(cr, COMPLETION_PARTIAL, COMPLETION_ALREADY)) {
+                                /* If this worked only partially, or if the user hit TAB even though we were
+                                 * complete already, then show the full list. */
+                                fputc('\n', stdout);
+
+                                STRV_FOREACH(nc, completions) {
+                                        const char *e;
+
+                                        e = startswith(*nc, strempty(string));
+                                        if (!e)
+                                                continue;
+
+                                        printf("\t%s%.*s%s%s\n",
+                                               ansi_grey(),
+                                               (int) (e - *nc), *nc,
+                                               ansi_normal(),
+                                               e);
+                                }
+
+                                /* Show the prompt again */
+                                fputs(ansi_highlight(), stdout);
+                                va_start(ap, text);
+                                vprintf(text, ap);
+                                va_end(ap);
+                                fputs(ansi_normal(), stdout);
+                                fputs(string, stdout);
+                        }
+
+                } else if (IN_SET(c, '\b', 127)) {
+                        /* Backspace */
+
+                        if (n == 0)
+                                fputc('\a', stdout); /* BEL */
+                        else {
+                                size_t m = utf8_last_length(string, n);
+
+                                char *e = string + n - m;
+                                clear_by_backspace(utf8_console_width(e));
+
+                                *e = 0;
+                                n -= m;
+                        }
+
+                } else if (c == 21) {
+                        /* Ctrl-u → erase all input */
+
+                        clear_by_backspace(utf8_console_width(string));
+                        string[n = 0] = 0;
+
+                } else if (c == 4) {
+                        /* Ctrl-d → cancel this field input */
+
+                        r = -ECANCELED;
```

Comment:
If you want to make this an error, we should actually handle it properly in the callers of this stuff. Currently we get an error which interrupts `homectl firstboot` and it doesn't ask any further questions.

Reply:
But that's the point of it, this should be a way to safely cancel the whole thing. askpw works the same way. This is intended behaviour really: it's not a clean skip, but a "fuck this" that fails the thing.

Reply:
a clean skip we already have: just hit enter. this is a failure operation however. And i think it's already handled the right way everywhere: the error bubbles up and aborts the program.